### DISCUSSION TITLE
Add new recipe to build casa environment with additional packages required by external toolboxes

### DISCUSSION THREAD

**Discussion by nsouedet** [brainvisa/brainvisa.github.io#98] [Q&A]
In my case I need to add tensorflow-gpu to be able to run an external implementation of efficientunet. To be able to install it, I need to install many dependencies provided by external NVidia repository. There is at least 3 possibilities to reach that goal : 

1. Create a casa-distro compatible recipe in an external project and modify casa-distro to be able to use this external recipe. This solution has the advantage to allow recipe to build both singularity, docker and virtualbox environments.
2. Modify casa-distro to be able to create a writable image. Changes in packages are done manually. It is a quick solution but it has the drawbacks to not be scripted and to not allow to build on docker and virtualbox environments directly.
3. Create a simple singularity recipe that uses casa-dev image has basis. The drawback here is that it does not allow to build on docker and virtualbox environments directly.

Which of these solutions should we use ?

**Comment by sapetnioc** [comment]
I prefer a solution compatible with Singularity and VirtualBox. First, image modules must be defined. These modules are chainable image modification recipes. They could be defined inside the Capsul project (accessible using a simple name for the user) or eventually in an external project. I am not found of modules not included in the project because they require more work for their specification and may be an easy entry point for Trojan horses (to date building recipes are Python code that will have to be downloaded and executed).
We could add a command to customize an image. This command would create a writable image and give the possibility to add modules (but could also be used to create a simple writable image). For instance:
```
casa_distro custom_image add=pytorch
```
The image would simply be tagged as customized in metadata and could not be upgraded anymore.
We should also be able to list available modules (with a short description ?):
```
casa_distro custom_image -hi
```
 
If we will probably have to double think about the user API because of the differences between Singularity and VirtualBox. In Singularity we have only image files. In VirtualBox, we have not only images files that cannot be modified and virtual machines that are the result of the importation of an image in VirtualBox and can be modified. What is the expected behavior in the two cases ? It is not yet very clear to me.

**Comment by nsouedet** [comment]
Many thanks, Yann, for your answers. 

1. Do you think the casa_distro command should create the custom image ?
I think that it should be the casa_distro_admin command because all other image creations are made by casa_distro_admin as it supposes to have root accesses. 

2. I think that adding modules to existing images is the ideal way to create custom images but it will suppose to manage dependencies and I am not sure that we want to be a new "system" with a specific dependency manager. I think that we just want to be able to rebuild 
a custom image in many formats (docker, singularity, virtualbox) from a script and/or a configuration file that enables to run an external toolbox.  And casa_distro will have search for a casa-dev-<toolbox> recipe in the image-recipes directories of projects in the dev environment. The advantage of this is that each external project is responsible of its image distribution, the main drawback is that there is not mutualisation between toolboxes and that a user should get multiple container file to run different toolboxes ... I think that a new image specific for a toolbox may be done using a command like :
```bash
casa_distro_admin create_base_image type="dev" base="casa-dev-5.0-8.sif" name="casa-dev-<toolbox>-5.0-8.sif"
```
